File: Interview Problems/Parking Lot/Java Code/service/SlotService.java
package service;

import domain.ParkingSlot;
import domain.Vehicle;
import repository.SlotRepository;

import java.util.Optional;
import java.util.UUID;

public class SlotService {

    private final SlotRepository slotRepository;

    public SlotService(SlotRepository slotRepository) {
        this.slotRepository = slotRepository;
        System.out.println("[SERVICE] SlotService initialized");
    }

    /**
     * Finds and marks as occupied the first available slot matching the vehicle type.
     * NOTE: In production, the find + occupy must be a single atomic DB transaction
     * (SELECT FOR UPDATE) to prevent concurrent threads picking the same slot.
     */
    public ParkingSlot allocateSlot(Vehicle.VehicleType vehicleType) {
        System.out.println("[SERVICE] Allocating slot for vehicle type: " + vehicleType);

        Optional<ParkingSlot> availableSlot = slotRepository.findAvailableSlot(vehicleType);
        if (availableSlot.isEmpty()) {
            throw new IllegalStateException("No available slot for vehicle type: " + vehicleType);
        }

        ParkingSlot slot = availableSlot.get();
        slot.occupy();
        slotRepository.update(slot);

        System.out.println("[SERVICE] Slot allocated: " + slot.getId() + " on floor " + slot.getFloorNumber());
        return slot;
    }

    public void releaseSlot(UUID slotId) {
        System.out.println("[SERVICE] Releasing slot: " + slotId);

        ParkingSlot slot = slotRepository.findById(slotId)
                .orElseThrow(() -> new IllegalArgumentException("Slot not found: " + slotId));

        slot.release();
        slotRepository.update(slot);
        System.out.println("[SERVICE] Slot released: " + slotId);
    }
}

File: Interview Problems/Parking Lot/Python Code/service/slot_service.py
import uuid
from domain.vehicle import VehicleType
from domain.parking_slot import ParkingSlot
from repository.slot_repository import SlotRepository

class SlotService:
    def __init__(self, slot_repo: SlotRepository):
        self.slot_repository = slot_repo
        print("[SERVICE] SlotService initialized")

    def allocate_slot(self, vehicle_type: VehicleType) -> ParkingSlot:
        print(f"[SERVICE] Allocating slot for vehicle type: {vehicle_type.name}")
        available_slot = self.slot_repository.find_available_slot(vehicle_type)
        if not available_slot:
            raise RuntimeError(f"No available slot for vehicle type: {vehicle_type.name}")

        available_slot.occupy()
        self.slot_repository.update(available_slot)
        print(f"[SERVICE] Slot allocated: {available_slot.id} on floor {available_slot.floor_number}")
        return available_slot

    def release_slot(self, slot_id: uuid.UUID):
        print(f"[SERVICE] Releasing slot: {slot_id}")
        slot = self.slot_repository.find_by_id(slot_id)
        if not slot:
            raise ValueError(f"Slot not found: {slot_id}")

        slot.release()
        self.slot_repository.update(slot)
        print(f"[SERVICE] Slot released: {slot_id}")
